### ISSUE TITLE
Question about how to set the max heap size when using "AnnotateBamWithUmis"

### ISSUE THREAD

**Issue by lxwgcool** [fulcrumgenomics/fgbio#595]
Question description:
I use UMI fastq file to annotate aligned BAM file through "fgbio AnnotateBamWithUmis" in our cluster.  
I keep getting error "OutOfMemoryError: GC Overhead Limit Exceeded" until I increased the maximum heap size of JVM from 16G to 48G (java -Xmx48g -jar). 

Here is the size of these two input files (UMI fastq is 717M with gz format, and BAM is 13G): 
717M Jul 13 11:59 UMI.fastq.gz
13G    Jul 13 11:59 HQ_paired_dedup_nophix.bam

I was so surprised that 16G heap size was failed to handle these two inputs. 
In order to let our pipeline can be ran smoothly without error, can you give me a suggestion for the question below: how to dynamic estimate/set the maximum size of heap for JVM according to the size of input files (UMI fastq + BAM)? 

System Info: 
CentOS release 6.10 (Final)
Total MEM: 220.8G
Number of Cores: 40 

Java Info:
openjdk 15-ea 2020-09-15
OpenJDK Runtime Environment (build 15-ea+4-63)
OpenJDK 64-Bit Server VM (build 15-ea+4-63, mixed mode, sharing)

Thanks so much for your help

**Comment by nh13** [comment]
@lxwgcool did you see this issue and the comments?  https://github.com/fulcrumgenomics/fgbio/issues/363.

**Comment by nh13** [comment]
I suppose if you could guarantee the reads were in the same order in the BAM and FASTQ, we wouldn't need to load all FASTQ into memory.  But then this tool would be rather trivial.

**Comment by lxwgcool** [comment]
@nh13 
Thank you so much for prompt reply, I just reviewed the thread you mentioned. For my case, I think using "AnnotateBamWithUmis" may be my only choice. This is because our goal is using UMI fastq reads to annotate many existing aligned BAM files (generate RX tag). I have several questions below:   

Question 1: May I know how to make reads in BAM and UMI fastq be the same order? 
Here are the key steps about how our BAM file be generated 
(1) bcl2fastq, -> PE reads
(2) Trimmomatic 
(3) Alignment by using  NovoAlign  
(4) Remove PCR duplicate and Phix reads by using Picard
(5) samtools sort

Question 2: For my understanding, making the reads in BAM and UMI fastq with the same order is only a suggestion for developer to change the source code to let fgbio be more efficiency,  In another word, if we download fgbio from github and run it directly, it will always load all fastq file into RAM no matter whether or not the reads order in BAM is consistent with the order in UMI fastq. 
Please correct me if I was wrong. 

Question 3: Finally, even the fastq be fully loaded into RAM to create hash table, may I know why 16G still does not enough? The size of my UMI fastq file is only 717 M with gz format. In another words, if we decompress it, the size of such UMI fastq reads won't larger than 8G. Why it still reports OOM even I set the maximum size of heap to 16G? 

Thanks so much for  your help
Best
Xin

**Comment by nh13** [comment]
@lxwgcool I think you want to change your upstream process if you can (see @tfenne's [comment about this ](https://github.com/fulcrumgenomics/fgbio/issues/363#issuecomment-363967470)).   Another thought I had was to split the UMI FASTQ up into smaller chunks, and run the tool multiple times on each chunk.  It's inefficient, but it will work right now.

> Question 1: May I know how to make reads in BAM and UMI fastq be the same order?

I would sort by read name.  For BAM, you could use `samtools sort -n`.  For FASTQ, it should be pretty easy to implement on your end.  But even if you had them sorted as such, we'd need to update the tool to not load all data into RAM.  At present, we are prioritizing (paid) client work, so that might be something that would take a while if you want us to do it.

Also, why don't you just extract UMIs during the demultiplexing (`bcl2fastq`) step?  Do you have control over that step?

> Question 2: For my understanding, making the reads in BAM and UMI fastq with the same order is only a suggestion for developer to change the source code to let fgbio be more efficiency, In another word, if we download fgbio from github and run it directly, it will always load all fastq file into RAM no matter whether or not the reads order in BAM is consistent with the order in UMI fastq.

It really depends on the implementation we make.  For example, we could have a command line option that does the sort for you on both the FASTQs and BAM, then runs the tool.  Then you would run it the same way always.

> Question 3: Finally, even the fastq be fully loaded into RAM to create hash table, may I know why 16G still does not enough? The size of my UMI fastq file is only 717 M with gz format. In another words, if we decompress it, the size of such UMI fastq reads won't larger than 8G. Why it still reports OOM even I set the maximum size of heap to 16G?

FASTQs compress well, so 717M is likely 4-10x larger, then comes all the overhead of object allocation in memory.  It's an art more than a science.

# TL;DR

Let me think about this more if you don't mind.  If this is a priority for you, then [feel free to reach out to us](https://www.fulcrumgenomics.com/contact/).

**Comment by lxwgcool** [comment]
Thank you so much for your prompt reply @nh13 

In our current pipeline, we use "bcl2fastq" for demultiplexing. It will generate three fastq files, including PE reads and UMI reads. That's why we try to use "AnnotateBamWithUmis" to add UMI back to the aligned BAM file. My teammate used to run "IlluminaBasecallsToSam" for demultiplexing, however it was too slow compared with bcl2fastq. I just checked "IlluminaBasecallsToSam" website today, and I noticed that "IlluminaBasecallsToSam" supports multiple-threads! That's a really good news, since many picard functions do not support parallel computing. I would make a testing to see if its performance could be as same as bcl2fastq. 

For the memory usage, I think the idea of creating hash table for UMI reads is very similar as the method I implemented in my paper:
https://github.com/lxwgcool/CircMarker
https://github.com/lxwgcool/CircDBG
I convert sqeunce (length < 16) into int 32, which helps save RAM. That's why I used to think 16G is big enough the save the hash table for 717M UMI reads. 

Thanks so much for your comments. I think I am all set. 
Best
Xin

**Comment by nh13** [comment]
@lxwgcool for sure we could implement string compaction, and Java9 does that (see Tim's comment).  Feel free to re-open if this is still an issue.

**Comment by nh13** [comment]
Also, don't forget that we need to store the read name in RAM so we can have a map from read name to bases !

**Comment by lxwgcool** [comment]
@nh13 
I am asking our HPC admin to install Java9. 
May I know the Java 9 you mentioned is for openjdk java or Oracle Java, or does it matter?
Thanks again
Best
Xin
